internal/pkg/es/client.go

```diff
@@ -6,21 +6,60 @@ package es
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"net/http"
 	"runtime"
+	"sync"
+	"syscall"
+	"time"
 
 	"go.elastic.co/apm/module/apmelasticsearch/v2"
 
 	"github.com/elastic/fleet-server/v7/internal/pkg/build"
 	"github.com/elastic/fleet-server/v7/internal/pkg/config"
 	"github.com/rs/zerolog"
 
+	backoff "github.com/cenkalti/backoff/v4"
 	"github.com/elastic/go-elasticsearch/v8"
 )
 
+const (
+	initialRetryBackoff = 500 * time.Millisecond
+	maxRetryBackoff     = 10 * time.Second
+	randomizationFactor = 0.5
+	defaultMaxRetries   = 5
+)
+
 type ConfigOption func(config *elasticsearch.Config)
 
+var retriableErrors []error
+var retriableErrorsLock sync.RWMutex
+
+func applyDefaultOptions(escfg *elasticsearch.Config) {
+	exp := backoff.NewExponentialBackOff()
+	exp.InitialInterval = initialRetryBackoff
+	exp.RandomizationFactor = randomizationFactor
+	exp.MaxInterval = maxRetryBackoff
+
+	opts := []ConfigOption{
+		WithRetryOnErr(syscall.ECONNREFUSED), // server not ready
+		WithRetryOnErr(syscall.ECONNRESET),   // server may be restarting
+
+		WithRetryOnStatus(http.StatusTooManyRequests),
+		WithRetryOnStatus(http.StatusBadGateway),
+		WithRetryOnStatus(http.StatusServiceUnavailable),
+		WithRetryOnStatus(http.StatusGatewayTimeout),
```

Comment:
If we want to keep to HTTP semantics, we could also retry 408 and 425. A cursory search showed that ES can return 408 in certain circumstances - notably when the client sets a timeout for the request, and ES is unable to process the request in that time. Not sure if that applies to our use in Fleet Server or not.